proposals/threads/DataSegmentInitialization.md

```diff
@@ -0,0 +1,177 @@
+# Proposal to conditionally initialize segments
+
+This page describes a proposal for providing a mechanism to skip data or
+element segment initialization when instantiating a module.
+
+Although the following rationale applies only to data segments, this proposal
+suggests that the proposed solutions apply to element segments as well for
+consistency.
+
+## Rationale
+
+Under the current threading proposal, to share a module between multiple
+agents, the module must be instantiated multiple times: once per agent.
+Instantiation initializes linear memory with the contents in the module's data
+segments. If the memory is shared between multiple agents, it will be
+initialized multiple times, potentially overwriting stores that occurred after
+the previous initializations.
+
+For example:
+
+```webassembly
+// The module.
+(module
+  (memory (export "memory") 1)
+
+  ;; Some value used as a counter.
+  (data (i32.const 1) "\0")
```

Comment:
Is there should be `i32.const 0`? IIUC, `addOne` expects counter to be at 0.

Reply:
Oops, thanks!

Reply:
@binji fixed https://github.com/WebAssembly/threads/pull/66 )